wrapper/src/main/java/software/amazon/jdbc/dialect/DialectManager.java

```diff
@@ -0,0 +1,210 @@
+      final Dialect userDialect = knownDialectsByCode.get(knownEndpointDialectCode);
+      if (userDialect != null) {
+        this.dialectCode = knownEndpointDialectCode;
+        this.dialect = userDialect;
+        return userDialect;
+      } else {
+        throw new SQLException(
+            Messages.get("DialectManager.unknownDialectCode", new Object[] {knownEndpointDialectCode}));
+      }
+    }
+
+    if (StringUtils.isNullOrEmpty(driverProtocol)) {
+      throw new IllegalArgumentException("protocol");
+    }
+
+    String host = url;
+    final List<HostSpec> hosts = this.connectionUrlParser.getHostsFromConnectionUrl(url, true);
+    if (!Utils.isNullOrEmpty(hosts)) {
+      host = hosts.get(0).getHost();
+    }
+
+    if (driverProtocol.contains("mysql")) {
+      this.canUpdate = true;
+      if (this.rdsHelper.isRdsDns(host)) {
+        this.dialectCode = DialectCodes.RDS_MYSQL;
+        this.dialect = knownDialectsByCode.get(DialectCodes.RDS_MYSQL);
+        return this.dialect;
+      }
+      this.dialectCode = DialectCodes.MYSQL;
+      this.dialect = knownDialectsByCode.get(DialectCodes.MYSQL);
+      return this.dialect;
+    }
+
+    if (driverProtocol.contains("postgresql")) {
+      this.canUpdate = true;
+      if (this.rdsHelper.isRdsDns(host)) {
+        this.dialectCode = DialectCodes.RDS_PG;
+        this.dialect = knownDialectsByCode.get(DialectCodes.RDS_PG);
+        return this.dialect;
+      }
+      this.dialectCode = DialectCodes.PG;
+      this.dialect = knownDialectsByCode.get(DialectCodes.PG);
+      return this.dialect;
+    }
+
+    if (driverProtocol.contains("mariadb")) {
+      this.canUpdate = true;
+      this.dialectCode = DialectCodes.MARIADB;
+      this.dialect = knownDialectsByCode.get(DialectCodes.MARIADB);
+      return this.dialect;
+    }
+
+    this.canUpdate = true;
+    this.dialectCode = DialectCodes.UNKNOWN;
+    this.dialect = knownDialectsByCode.get(DialectCodes.UNKNOWN);
+    return this.dialect;
+  }
+
+  public Dialect getDialect(
+      final @NonNull String originalUrl,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Connection connection) throws SQLException {
+
+    if (!this.canUpdate) {
+      return this.dialect;
+    }
+
+    final List<String> dialectCandidates = this.dialect.getDialectUpdateCandidates();
+    if (dialectCandidates != null) {
+      for (String dialectCandidateCode : dialectCandidates) {
+        Dialect dialectCandidate = knownDialectsByCode.get(dialectCandidateCode);
+        if (dialectCandidate == null) {
+          throw new SQLException(
+              Messages.get("DialectManager.unknownDialectCode", new Object[] {dialectCandidateCode}));
+        }
+        boolean isDialect = dialectCandidate.isDialect(connection);
+        if (isDialect) {
+          this.canUpdate = false;
+          this.dialectCode = dialectCandidateCode;
+          this.dialect = dialectCandidate;
```

Comment:
These three are always updated together, can we create a util method updating all three at once to avoid potential human mistakes